

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Macromolecular models &mdash; MacSyFinder 2.0 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="MacSyFinder’s functioning" href="functioning.html" />
    <link rel="prev" title="Output format" href="outputs.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home" alt="Documentation Home"> MacSyFinder
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">User Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="index.html#running-macsyfinder">Running MacSyFinder</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html#macsyfinder-functioning">MacSyFinder functioning</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Macromolecular models</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#structure-of-a-macsy-model-package">Structure of a macsy-model package</a></li>
<li class="toctree-l4"><a class="reference internal" href="#principles-and-how-to-write-macsy-models-definitions">Principles, and how to write macsy-models definitions</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#installing-and-sharing-models">Installing and sharing models</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#how-to-install-new-models">How to install new models</a></li>
<li class="toctree-l4"><a class="reference internal" href="#where-the-models-are-located">Where the models are located</a></li>
<li class="toctree-l4"><a class="reference internal" href="#writing-my-own-macsy-model-package">Writing my own macsy-model package</a></li>
<li class="toctree-l4"><a class="reference internal" href="#share-your-models">Share your models</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="functioning.html">MacSyFinder’s functioning</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/index.html">Developer Guide</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MacSyFinder</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="index.html">User Guide</a> &raquo;</li>
        
      <li>Macromolecular models</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/user_guide/model_definition.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="macromolecular-models">
<span id="model-definition"></span><h1>Macromolecular models<a class="headerlink" href="#macromolecular-models" title="Permalink to this headline">¶</a></h1>
<p>MacSyFinder relies on the definition of models of macromolecular systems as a <strong>set of models’ components</strong>
to be searched by similarity search, and a <strong>set of rules</strong> regarding their genomic organization and
their requirement level to make a complete system (mandatory, accessory components, number of components required).</p>
<p>See <a class="reference internal" href="#model-definition-grammar-label"><span class="std std-ref">below</span></a> for more details on MacSyFinder’s modelling scheme and the section
on <a class="reference internal" href="functioning.html#functioning"><span class="std std-ref">Functioning</span></a> for the principles of the MacSyFinder’s search engine.</p>
<p>A <strong>MacSyFinder model</strong> (macsy-model for short) is the association of several elements:</p>
<blockquote>
<div><ul class="simple">
<li><p>a <strong>definition</strong> which describes the system to detect with a specific <strong>XML grammar</strong> that is described <a class="reference internal" href="#model-definition-grammar-label"><span class="std std-ref">below</span></a>.</p></li>
<li><p>a set of <a class="reference internal" href="#provide-hmm-label"><span class="std std-ref">HMM profiles</span></a>  (one per component/gene in the model) to enable the similarity search of the systems’ components with the HMMER program.</p></li>
</ul>
</div></blockquote>
<p>The models are grouped by <em>family</em> possibly gathering <em>sub-families</em> (multiple levels allowed), for instance <em>Secretion</em>, <em>Cas-proteins</em>…
A set of models from a same family (coherent set) of systems to detect is called hereafter a <strong>macsy-model package</strong> <code class="docutils literal notranslate"><span class="pre">NEW</span> <span class="pre">in</span> <span class="pre">V2</span></code>.</p>
<div class="section" id="structure-of-a-macsy-model-package">
<span id="package-structure"></span><h2>Structure of a macsy-model package<a class="headerlink" href="#structure-of-a-macsy-model-package" title="Permalink to this headline">¶</a></h2>
<p>A macsy-model package follows the following structure</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">family_name</span>
    <span class="o">|</span><span class="n">_______</span> <span class="n">metadata</span><span class="o">.</span><span class="n">yml</span>
    <span class="o">|</span><span class="n">_______</span> <span class="n">LICENCE</span>
    <span class="o">|</span><span class="n">_______</span> <span class="n">README</span><span class="o">.</span><span class="n">md</span>
    <span class="o">|</span><span class="n">_______</span> <span class="n">definitions</span>
    <span class="o">|</span>            <span class="o">|</span><span class="n">________</span> <span class="n">model_1</span><span class="o">.</span><span class="n">xml</span>
    <span class="o">|</span>            <span class="o">|</span><span class="n">________</span> <span class="n">model_2</span><span class="o">.</span><span class="n">xml</span>
    <span class="o">|</span>            <span class="p">:</span>
    <span class="o">|</span>
    <span class="o">|</span><span class="n">_______</span> <span class="n">profiles</span>
                 <span class="o">|</span><span class="n">________</span> <span class="n">geneA</span><span class="o">.</span><span class="n">hmm</span>
                 <span class="o">|</span><span class="n">________</span> <span class="n">geneB</span><span class="o">.</span><span class="n">hmm</span>
</pre></div>
</div>
<p>If the package contains sub-families</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">family_name</span>
    <span class="o">|</span><span class="n">_______</span> <span class="n">metadata</span><span class="o">.</span><span class="n">yml</span>
    <span class="o">|</span><span class="n">_______</span> <span class="n">LICENCE</span>
    <span class="o">|</span><span class="n">_______</span> <span class="n">README</span><span class="o">.</span><span class="n">md</span>
    <span class="o">|</span><span class="n">_______</span> <span class="n">definitions</span>
    <span class="o">|</span>            <span class="o">|</span><span class="n">________</span> <span class="n">subfamilyA</span>
    <span class="o">|</span>            <span class="o">|</span>            <span class="o">|</span><span class="n">________</span> <span class="n">model_1</span><span class="o">.</span><span class="n">xml</span>
    <span class="o">|</span>            <span class="o">|</span>            <span class="o">|</span><span class="n">________</span> <span class="n">model_2</span><span class="o">.</span><span class="n">xml</span>
    <span class="o">|</span>            <span class="o">|</span>
    <span class="o">|</span>            <span class="o">|</span><span class="n">________</span> <span class="n">subfamilyB</span>
    <span class="o">|</span>            <span class="o">|</span>            <span class="o">|</span><span class="n">________</span> <span class="n">model_3</span><span class="o">.</span><span class="n">xml</span>
    <span class="o">|</span>            <span class="o">|</span>            <span class="o">|</span><span class="n">________</span> <span class="n">model_4</span><span class="o">.</span><span class="n">xml</span>
    <span class="o">|</span>            <span class="o">|</span>
    <span class="o">|</span>            <span class="p">:</span>
    <span class="o">|</span>
    <span class="o">|</span><span class="n">_______</span> <span class="n">profiles</span>
                 <span class="o">|</span><span class="n">________</span> <span class="n">geneA</span><span class="o">.</span><span class="n">hmm</span>
                 <span class="o">|</span><span class="n">________</span> <span class="n">geneB</span><span class="o">.</span><span class="n">hmm</span>
</pre></div>
</div>
<p>For examples of macsy-model packages, please visit <a class="reference external" href="https://github.com/macsy-models">https://github.com/macsy-models</a></p>
</div>
<div class="section" id="principles-and-how-to-write-macsy-models-definitions">
<span id="writing-models"></span><h2>Principles, and how to write macsy-models definitions<a class="headerlink" href="#principles-and-how-to-write-macsy-models-definitions" title="Permalink to this headline">¶</a></h2>
<p>Macsy-models are written as XML files, and should be named with the name of the system to detect as a prefix,
and the XML file extension as a suffix. For example, ‘T1SS.xml’ for T1SS (Type I Secretion System).</p>
<p>A macsy-model defines a macromolecular System as:</p>
<ul class="simple">
<li><p>A set of <strong>components</strong> (<em>i.e.</em> proteins, or protein-coding genes given the context) with different attributes that are used for system’s <strong>content description</strong>.</p></li>
<li><p>Features regarding the <strong>genomic architecture</strong> of the systems’ components for system detection.</p></li>
<li><p>Rules for <strong>quorum</strong> specifying how many components are required to infer the presence of a complete system.</p></li>
</ul>
<div class="section" id="macsy-model-components">
<span id="components"></span><h3>Macsy-model Components<a class="headerlink" href="#macsy-model-components" title="Permalink to this headline">¶</a></h3>
<p>Four distinct <strong>types of components</strong> can be used to model the System’s content.
Components correspond to Gene objects in MacSyFinder’s implementation, and point to corresponding HMM protein profiles.</p>
<ul class="simple">
<li><p><strong>mandatory</strong> components represent components that are essential to be found to infer the system’s presence.</p></li>
<li><p><strong>accessory</strong> components correspond to components that can be found in some systems’ occurrence
(or quickly evolving components that are hard to detect with a single HMM profile and thus can be missed along similarity search).</p></li>
<li><p><strong>neutral</strong> components are used to build/extend clusters of proximal genes/components on the replicon analysed, but are not part of the quorum (i.e., not taken into account to assess the system’s presence). <code class="docutils literal notranslate"><span class="pre">NEW</span> <span class="pre">in</span> <span class="pre">V2</span></code></p></li>
<li><p><strong>forbidden</strong> components are components which presence is eliminatory for the system’s presence assessment.</p></li>
</ul>
</div>
<div class="section" id="specifying-a-genomic-organization">
<span id="model-definition-genomic-orga"></span><h3>Specifying a genomic organization<a class="headerlink" href="#specifying-a-genomic-organization" title="Permalink to this headline">¶</a></h3>
<p>Beyond its list of Components, a MacSyFinder’s model of a System is defined by the genomic organization of its components.
This genomic organization can be defined in several ways:</p>
<ul>
<li><p>the general System’s architecture, whether it is <cite>single-locus</cite> or <cite>multi-loci</cite> (encoded at one or several loci)</p></li>
<li><p>the co-localization criteria defined either at the System level or at the Gene (component) level:</p>
<blockquote>
<div><ul class="simple">
<li><p>the <cite>inter-gene-max-space</cite> parameter (system- or gene- wise)</p></li>
<li><p>the <cite>loner</cite> parameter (gene- wise)</p></li>
</ul>
</div></blockquote>
</li>
</ul>
<p>See <a class="reference internal" href="#model-definition-grammar-label"><span class="std std-ref">below</span></a> for more details on how to specify these parameters in a macsy-model.</p>
</div>
<div class="section" id="the-xml-hierarchy">
<span id="model-definition-grammar-label"></span><h3>The XML hierarchy<a class="headerlink" href="#the-xml-hierarchy" title="Permalink to this headline">¶</a></h3>
<p>A System’s model is defined using a specific XML grammar that is hereby described.
It consists in a hierarchic view of a Model that has specific features described through parameters, and is made of a set of Genes that have specific features themselves.
All these elements and corresponding parameters will parametrize the search of Systems matching the search by MacSyFinder, in terms of Gene content and genomic architecture criteria.</p>
<a class="reference internal image-reference" href="../_images/MSF_modelling.svg"><img alt="../_images/MSF_modelling.svg" class="align-left" height="1000px" src="../_images/MSF_modelling.svg" /></a>
<ul>
<li><p>The element root of a System’s model is “model”.</p>
<ul>
<li><p>It has a mandatory attribute: “inter_gene_max_space”, an integer representing the maximal number of components
without a match between two components with a match for a component profile in order to consider them contiguous (part of a same <em>Cluster</em>).</p></li>
<li><p>The version of the XML grammar (the actual version is “2.0”)</p></li>
<li><p>The element “model” may have attributes:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>min_mandatory_genes_required</strong>: an <em>integer</em> representing the minimal number of mandatory genes required
to infer the system’s presence.</p></li>
<li><p><strong>min_genes_required</strong>: an <em>integer</em> representing the minimal number of mandatory or accessory genes
(whose corresponding proteins match a profile of the model) required to infer the system’s presence.</p></li>
<li><p><strong>multi_loci</strong>: a <em>boolean</em> set to True (“1”, “true” or “True”) to allow the definition of “scattered” systems
(i.e., systems encoded at different genomic loci or by different gene <em>clusters</em>). If not specified, <em>default value is false</em>.</p></li>
</ul>
</div></blockquote>
</li>
<li><p>The model contains one or more element(s) “gene” that correspond(s) to the genetic components of the macromolecular system.</p></li>
</ul>
</li>
<li><p>The element “gene” has several mandatory attributes:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>name</strong>: a <em>string</em> representing the name of the component/gene which must match that of a profile enclosed in the profile directory of the macsy-model package (see <a class="reference internal" href="#provide-hmm-label"><span class="std std-ref">below</span></a>).</p></li>
<li><p><strong>presence</strong>: a <em>string</em> representing the status of the gene’s presence in the system. It can take four values among “mandatory”, “accessory”, “neutral”, “forbidden” (see above).</p></li>
</ul>
</div></blockquote>
</li>
</ul>
<blockquote>
<div><p>The element “gene” may have other attributes:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>loner</strong>: a <em>boolean</em>. A <em>loner</em> gene can be isolated on the genome and does not have to be part of a cluster of genes to be considered for system’s assessment ( <em>default false</em> ).</p></li>
<li><p><strong>multi_system</strong>: a <em>boolean</em>. If a gene has the feature “multi_system” (value set to “1”, “true” or “True”),
it means that it can be used to fill multiple systems’ occurrences - and thus be considered part of several systems. ( <em>default false</em> ).</p></li>
<li><p><strong>inter_gene_max_space</strong>: an <em>integer</em> that defines gene-wise value of system’s “inter_gene_max_space” parameter (see above). It supersedes the system-wise parameter to give the gene a specific co-localization parameter.</p></li>
</ul>
</div></blockquote>
<p>The element “gene” may have one “exchangeables” child element:</p>
<blockquote>
<div><ul class="simple">
<li><p>The element “exchangeables” can contain one or more elements “gene”.</p></li>
</ul>
<p>For a Gene to have “exchangeables” Genes listed, means that this Gene can be replaced <em>in the quorum</em> by the listed child Genes.</p>
</div></blockquote>
</div></blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If not specified by the user, several features will have their values assigned <strong>by default</strong>:</p>
<ul class="simple">
<li><p>the <strong>genomic architecture</strong> of the System being searched will consist in a <strong>single locus</strong>. If a System may be made of Genes from multiple loci, consider setting the <cite>multi_loci</cite> parameter to <cite>True</cite>.</p></li>
<li><p>the <strong>quorum parameters</strong> <cite>min_mandatory_genes_required</cite> and <cite>min_genes_required</cite> will be set to the number of mandatory Genes listed - the <cite>accessory</cite> Genes being deemed not required to infer a complete System.</p></li>
</ul>
</div>
<p>Example of a macsy-model definition in XML:</p>
<div class="highlight-xml notranslate"><div class="highlight"><pre><span></span><span class="nt">&lt;model</span> <span class="na">inter_gene_max_space=</span><span class="s">&quot;5&quot;</span> <span class="na">ver=</span><span class="s">&quot;2.0&quot;</span><span class="nt">&gt;</span>
  <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;gspD&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">&gt;</span>
     <span class="nt">&lt;exchangeables&gt;</span>
         <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;sctC&quot;</span><span class="nt">/&gt;</span>
     <span class="nt">&lt;/exchangeables&gt;</span>
  <span class="nt">&lt;/gene&gt;</span>
  <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;sctN_FLG&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span> <span class="na">loner=</span><span class="s">&quot;1&quot;</span><span class="nt">/&gt;</span>
     <span class="nt">&lt;exchangeables&gt;</span>
         <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;gspE&quot;</span><span class="nt">/&gt;</span>
         <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;pilT&quot;</span><span class="nt">/&gt;</span>
     <span class="nt">&lt;/exchangeables&gt;</span>
  <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;sctV_FLG&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
  <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;flp&quot;</span> <span class="na">presence=</span><span class="s">&quot;accessory&quot;</span><span class="nt">/&gt;</span>
<span class="nt">&lt;/model&gt;</span>
</pre></div>
</div>
<p>In this example, the described System consists of three mandatory and one accessory components:</p>
<blockquote>
<div><ul class="simple">
<li><p>Two components, the Gene “GspD” and the Gene “sctN_FLG” can respectively be replaced by sctC, and gspE and pilT genes in the quorum.</p></li>
<li><p>To be considered as part of such System, the components should be co-localized in loci (Clusters of Genes), which in this case would amount to being located from each other at a distance of 5-Genes maximum, except for the Gene “sctN_FLG” that is allowed to be located “alone” in the genome being investigated, by a <cite>loner</cite> parameter being set to True. As the <cite>multi_loci</cite> parameter is not set, by default the System should be made of a single locus (Cluster of co-localized Genes - except for the ones listed as <cite>loners</cite>).</p></li>
<li><p>To be considered a complete System, the quorum of Genes should be reached. In this case, the <cite>min_genes_required</cite> and <cite>min_mandatory_genes_required</cite> are not specified and therefore assigned to their default values: <cite>min_mandatory_genes_required</cite> is set to the number of mandatory Genes listed as well as the <cite>min_genes_required</cite> parameter (see above).</p></li>
</ul>
</div></blockquote>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<ul class="simple">
<li><p>a gene is identified by its name.</p></li>
<li><p>this name is case sensitive.</p></li>
<li><p>this name must be unique inside a family of models.</p></li>
<li><p>a HMM profile with a gene-based name must exist in the <cite>profiles</cite> directory of the macsy-model package (see <a class="reference internal" href="#provide-hmm-label"><span class="std std-ref">below</span></a>).</p></li>
</ul>
</div>
</div>
<div class="section" id="providing-hmm-profiles">
<span id="provide-hmm-label"></span><h3>Providing HMM profiles<a class="headerlink" href="#providing-hmm-profiles" title="Permalink to this headline">¶</a></h3>
<p>For each gene mentioned in each model you have to provide <strong>a HMM profile</strong>
to enable the similarity search of this gene. The HMM profile must have been created by the user from a curated multiple sequence alignment with the <cite>hmmbuild</cite> program
from the <a class="reference external" href="http://hmmer.org/">HMMER package</a>, or can have been obtained from HMM profiles’ databases such as <a class="reference external" href="https://dx.doi.org/10.1093%2Fnar%2Fgkg128">TIGRFAM</a> or <a class="reference external" href="https://pfam.xfam.org/">PFAM</a> .</p>
<p>This profile <em>MUST</em> have the same name as the name of the gene mentioned in the definition.
For instance, a component named “GeneA” in the macsy-model would correspond by default to a HMM profile “GeneA.hmm” enclosed in the macsy-model package.
The names are <strong>case-sensitive</strong>. All HMM profiles must be placed in the <cite>profiles</cite> directory of the macsy-model package.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For a detailed tutorial on how to define your macsy-model’s features, parameters and HMM profiles, you can have a look at our cookbook in <a class="reference external" href="https://link.springer.com/protocol/10.1007/978-1-4939-7033-9_1">this book chapter</a> .</p>
</div>
</div>
</div>
</div>
<div class="section" id="installing-and-sharing-models">
<h1>Installing and sharing models<a class="headerlink" href="#installing-and-sharing-models" title="Permalink to this headline">¶</a></h1>
<div class="section" id="how-to-install-new-models">
<h2>How to install new models<a class="headerlink" href="#how-to-install-new-models" title="Permalink to this headline">¶</a></h2>
<p>MacSyFinder does not provide models. You must install models before using it.
The <code class="docutils literal notranslate"><span class="pre">macsydata</span></code> utility tool is shipped with <cite>MacSyFinder</cite> to deal with macsy-models:</p>
<p>macsydata &lt;subcommand&gt; [options]</p>
<p>The main sub-commands are</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">macsydata</span> <span class="pre">available</span></code> to get the list of macsy-models available</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">macsydata</span> <span class="pre">search</span></code> to search a model given its name or a pattern in its description</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">macsydata</span> <span class="pre">install</span></code> to install a macsy-model package (the installed version can be set see –help)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">macsydata</span> <span class="pre">cite</span></code> to retrieve information on how to cite the model</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">macsydata</span> <span class="pre">--help</span></code> to get the extended list of available subcommands</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">macsydata</span> <span class="pre">&lt;subcommand&gt;</span> <span class="pre">--help</span></code> to get help about the specified subcommand</p></li>
</ul>
<p><em>macsydata</em> is <code class="docutils literal notranslate"><span class="pre">NEW</span> <span class="pre">in</span> <span class="pre">V2</span></code></p>
</div>
<div class="section" id="where-the-models-are-located">
<h2>Where the models are located<a class="headerlink" href="#where-the-models-are-located" title="Permalink to this headline">¶</a></h2>
<p>MacSyFinder looks at several locations to find macsy-models.</p>
<div class="section" id="system-wide-installation">
<h3>system-wide installation<a class="headerlink" href="#system-wide-installation" title="Permalink to this headline">¶</a></h3>
<p>By default <em>macsydata</em> installs models in a shared location (set by –install-data option) that is
<cite>/usr/share/macsyfinder/</cite> or <cite>/usr/local/share/macsyfinder</cite> depending on your Operating System distribution.
If you use a <em>virtualenv</em>, the shared resources are located in the <cite>&lt;virtualenv&gt;/share/macsyfinder</cite> directory.</p>
</div>
<div class="section" id="user-wide-installation">
<h3>user-wide installation<a class="headerlink" href="#user-wide-installation" title="Permalink to this headline">¶</a></h3>
<p>If you don’t own rights to install system-wide, you can install models in the MacSyFinder’s cache
located in your home: <cite>$HOME/.macsyfinder/data/</cite>.
<em>macsydata</em> installs packages in this location when you use the <cite>–user</cite> option.
The packages installed in user land is added to the system-wide packages.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If two packages have the same name, the package in the user land supersedes the system-wide package.</p>
</div>
</div>
<div class="section" id="project-wide-installation">
<h3>project-wide installation<a class="headerlink" href="#project-wide-installation" title="Permalink to this headline">¶</a></h3>
<p>If you cannot install macsy-model packages in system or user land locations, you can specify a
specific location with the <code class="docutils literal notranslate"><span class="pre">--models-dir</span></code> <a class="reference internal" href="input.html#path-options"><span class="std std-ref">command-line option</span></a>. The path must point at a directory
that contains macsy-model packages as described <a class="reference internal" href="#package-structure"><span class="std std-ref">above</span></a>.</p>
<blockquote>
<div></div></blockquote>
</div>
</div>
<div class="section" id="writing-my-own-macsy-model-package">
<span id="model-package"></span><h2>Writing my own macsy-model package<a class="headerlink" href="#writing-my-own-macsy-model-package" title="Permalink to this headline">¶</a></h2>
<p>The whole package structure is described <a class="reference internal" href="#package-structure"><span class="std std-ref">above</span></a> and requires five different types of files described below to be complete:</p>
<ul class="simple">
<li><p>a metadata file</p></li>
<li><p>a README.md file</p></li>
<li><p>a LICENCE file</p></li>
<li><p>macsy-models definition(s)</p></li>
<li><p>HMM profiles</p></li>
</ul>
<div class="section" id="metadata-file">
<h3>metadata file<a class="headerlink" href="#metadata-file" title="Permalink to this headline">¶</a></h3>
<p>This file contains some meta information about the package itself.
It is in <a class="reference external" href="https://en.wikipedia.org/wiki/YAML">YAML</a> format and must have the following structure:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nn">---</span>
<span class="nt">maintainer</span><span class="p">:</span>
  <span class="nt">name</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">The name of the person who maintains/to contact for further information. (required)</span>
  <span class="nt">email</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">The email of the maintainer (required)</span>
<span class="nt">short_desc</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">A one line description of the package (can e.g. be used for *macsydata* searches) (required)</span>
<span class="nt">vers</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">The package version (required)</span>
<span class="nt">cite</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">The publication(s) to cite by the user when the package is used (optional, used by `macsydata cite`)</span>
<span class="nt">doc</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">Where to find extended documentation (optional)</span>
<span class="nt">licence</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">The licence under the package is released (optional but highly recommended)</span>
<span class="nt">copyright</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">The copyright of the package (optional)</span>
</pre></div>
</div>
<p>For example:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nn">---</span>
<span class="nt">maintainer</span><span class="p">:</span>
   <span class="nt">name</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">first name last name</span>
   <span class="nt">email</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">login@my_domain.com</span>
<span class="nt">short_desc</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">Models for 15 types of secretion systems or bacterial appendages (T1SS, T2SS, T3SS, T4P, pT4SSt, pT4SSi, T5aSS, T5bSS, T5bSS, T6SSi, T6SSii, T6SSiii, Flagellum, Tad, T9SS).</span>
<span class="nt">vers</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">0.0a1</span>
<span class="nt">cite</span><span class="p">:</span>
   <span class="p p-Indicator">-</span> <span class="p p-Indicator">|</span>
     <span class="no">Abby Sophie S., Cury Jean, Guglielmini Julien, Néron Bertrand, Touchon Marie, Rocha Eduardo P. C. (2016).</span>
     <span class="no">Identification of protein secretion systems in bacterial genomes.</span>
     <span class="no">In Scientific Reports, 6, pp. 23080.</span>
     <span class="no">http://dx.doi.org/10.1038/srep23080</span>
<span class="nt">doc</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">https://github.com/macsy-models/TXSS</span>
<span class="nt">licence</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">CC BY-NC-SA 4.0 (https://creativecommons.org/licenses/by-nc-sa/4.0/)</span>
<span class="nt">copyright</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">2014-2020, Institut Pasteur, CNRS</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This <cite>metadata.yml</cite> file is <strong>mandatory</strong>. Without this file your archive/repository will not be considered as a <em>macsy-model package</em>.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p><em>-</em> specify an item of yaml list</p></li>
<li><p><em>|</em> is used to specify a single item but over multiple lines.</p></li>
</ul>
</div>
</div>
<div class="section" id="readme-md">
<h3>README.md<a class="headerlink" href="#readme-md" title="Permalink to this headline">¶</a></h3>
<p>A description of the package: what kind of systems the package models, how to use it etc… in <a class="reference external" href="https://guides.github.com/features/mastering-markdown/">markdown</a> format.</p>
</div>
<div class="section" id="licence">
<h3>LICENCE<a class="headerlink" href="#licence" title="Permalink to this headline">¶</a></h3>
<p>The licence use to protect and share your work.
If you don’t know which licence to choose, have a look at <a class="reference external" href="https://creativecommons.org/share-your-work/">CreativeCommons</a>
<em>This file is optional, but highly recommended.</em></p>
</div>
</div>
<div class="section" id="share-your-models">
<h2>Share your models<a class="headerlink" href="#share-your-models" title="Permalink to this headline">¶</a></h2>
<p>If you want to share your models you can create a <a class="reference internal" href="#model-package"><span class="std std-ref">macsy-model package</span></a> in your github repository.</p>
<ol class="arabic simple">
<li><p>check the validity of your package with the <code class="docutils literal notranslate"><span class="pre">macsydata</span> <span class="pre">check</span></code> command.</p></li>
<li><p>create a tag, and submit a pull request to <a class="reference external" href="https://github.com/macsy-models">https://github.com/macsy-models</a> organization.</p></li>
<li><p>when your pull request (PR) is accepted, the model package becomes automatically available to the community through the <em>macsydata</em> tool.</p></li>
</ol>
<p>If you don’t want to submit a PR you can provide the tag release tarball (tar.gz) as is to your collaborators.
This archive will also be usable with the <cite>macsydata</cite> tool.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="functioning.html" class="btn btn-neutral float-right" title="MacSyFinder’s functioning" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="outputs.html" class="btn btn-neutral float-left" title="Output format" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2014-2020, &#39;Institut Pasteur (Paris), CNRS&#39;

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  

<a href="https://github.com/gem-pasteur/macsyfinder">
    <img style="position: absolute; top: 0; right: 0; border: 0;"
         src="https://camo.githubusercontent.com/38ef81f8aca64bb9a64448d0d70f1308ef5341ab/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f6769746875622f726962626f6e732f666f726b6d655f72696768745f6461726b626c75655f3132313632312e706e67"
         alt="Fork me on GitHub"
         data-canonical-src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png">
</a>



</body>
</html>