

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>ERVRegions &mdash; ERVsearch 1.0.0 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home" alt="Documentation Home"> ERVsearch
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="introduction.html">ERVsearch</a></li>
<li class="toctree-l1"><a class="reference internal" href="parameters.html">Parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="functions.html">Functions</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">ERVsearch</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>ERVRegions</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/pages/ERVregions.md.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="ervregions">
<h1>ERVRegions<a class="headerlink" href="#ervregions" title="Permalink to this headline">¶</a></h1>
<hr class="docutils" />
<div class="section" id="makecleanbeds">
<h2>makeCleanBeds<a class="headerlink" href="#makecleanbeds" title="Permalink to this headline">¶</a></h2>
<p><strong>Input Files</strong><br>
<code class="docutils literal notranslate"><span class="pre">grouped.dir/GENE_groups.tsv</span></code><br></p>
<p><strong>Output Files</strong><br>
<code class="docutils literal notranslate"><span class="pre">clean_beds.dir/GENE.bed</span></code><br></p>
<p><strong>Parameters</strong><br>
None<br></p>
<p>Generates a bed file for each gene which contains the co-ordinates of the ORFs which have passed all filtering criteria in the Screen section.</p>
</div>
<div class="section" id="makecleanfastas">
<h2>makeCleanFastas<a class="headerlink" href="#makecleanfastas" title="Permalink to this headline">¶</a></h2>
<p><strong>Input Files</strong><br>
<code class="docutils literal notranslate"><span class="pre">clean_beds.dir/GENE.bed</span></code><br>
<code class="docutils literal notranslate"><span class="pre">genome.fa</span></code><br></p>
<p><strong>Output Files</strong><br>
<code class="docutils literal notranslate"><span class="pre">clean_fastas.dir/GENE.fasta</span></code><br></p>
<p><strong>Parameters</strong><br>
None<br></p>
<p>Fasta files are generated containing the sequences of the regions listed by makeCleanBeds. These are extracted from the host chromosomes using bedtools getfasta (https://bedtools.readthedocs.io/en/latest/content/tools/getfasta.html).</p>
</div>
<div class="section" id="findervregions">
<h2>findERVRegions<a class="headerlink" href="#findervregions" title="Permalink to this headline">¶</a></h2>
<p><strong>Input Files</strong><br>
<code class="docutils literal notranslate"><span class="pre">clean_fastas.dir/*.fasta</span></code><br></p>
<p><strong>Output Files</strong><br>
<code class="docutils literal notranslate"><span class="pre">ERV_regions.dir/all_ORFs.bed</span></code><br>
<code class="docutils literal notranslate"><span class="pre">ERV_regions.dir/all_regions.bed</span></code><br>
<code class="docutils literal notranslate"><span class="pre">ERV_regions.dir/multi_gene_regions.bed</span></code><br>
<code class="docutils literal notranslate"><span class="pre">ERV_regions.dir/regions.fasta</span></code><br></p>
<p><strong>Parameters</strong><br>
<code class="docutils literal notranslate"><span class="pre">[regions]</span> <span class="pre">maxdist</span></code><br></p>
<p>Combines the files containng the ORF regions for the different retroviral genes and merges any regions which are within <code class="docutils literal notranslate"><span class="pre">regions_maxdist</span></code> of each other to find larger regions containing multiple genes.
The <code class="docutils literal notranslate"><span class="pre">all_ORFs.bed</span></code> output file is the concatenated and sorted bed files, <code class="docutils literal notranslate"><span class="pre">all_regions.bed</span> </code>contains the merged regions with any ORFs within regions_maxdist of each other (end to end) combined, plus all regions with a single ORF, generated from all_regions.bed using bedtools merge (https://bedtools.readthedocs.io/en/latest/content/tools/merge.html). The name, strand and score columns are concatenated for merged regions, delimited with a “,”.
<code class="docutils literal notranslate"><span class="pre">multi_gene_regions.bed</span></code> contains only the regions which were found to contain multiple ORFs, <code class="docutils literal notranslate"><span class="pre">regions.fasta</span></code> is the sequence of these regions in FASTA format. At this point this includes regions with multiple ORFs from the same gene (e.g. two <em>pol</em> ORFs).</p>
</div>
<div class="section" id="makeregiontables">
<h2>makeRegionTables<a class="headerlink" href="#makeregiontables" title="Permalink to this headline">¶</a></h2>
<p><strong>Input Files</strong><br>
<code class="docutils literal notranslate"><span class="pre">ERV_regions.dir/multi_gene_regions.bed</span></code><br>
<code class="docutils literal notranslate"><span class="pre">grouped.dir/*_groups.tsv</span></code><br>
<code class="docutils literal notranslate"><span class="pre">genome.fa</span></code><br></p>
<p><strong>Output Files</strong>
<code class="docutils literal notranslate"><span class="pre">ERV_regions.dir/ERV_regions_final.tsv</span></code><br>
<code class="docutils literal notranslate"><span class="pre">ERV_regions.dir/ERV_regions_final.bed</span></code><br>
<code class="docutils literal notranslate"><span class="pre">ERV_regions.dir/ERV_regions_final.fasta</span></code><br></p>
<p><strong>Parameters</strong><br>
None<br></p>
<p>Takes a merged bed file consisting of regions of the genome identified as having more than one ERV-like ORF, finds the regions within this file
which contain more than one different gene (e.g. gag and pol instead of two gag ORFs) and outputs a formatted table of information about these
regions.</p>
<p>The output table (<code class="docutils literal notranslate"><span class="pre">ERV_regions_final.tsv</span></code>) will usually have 37 columns:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">name</span></code> - the final ID of the ERV region - the genes found plus an integer  e.g. gag_pol_12<br></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">chrom</span></code> - chromosome<br></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">start</span></code> - start position of the ERV region<br></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">end</span></code> - end position of the ERV region<br></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">strand</span></code> - strand of the ERv region<br></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">genus</span></code> - genus of the ERV region, can be multiple genera delimted by “|” if different genes had different genera<br></p></li>
<li><p>for each gene screened for (usually gag, pol and env)<br></p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">GENE_name</span></code> - the names of the ORFs for this gene in this region<br></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">GENE_ID</span></code> - the original IDs of the ORFs for this gene in this region<br></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">GENE_start</span></code> - the start position of this gene in this region (genome co-ordinates)<br></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">GENE_relative_start</span></code> - the start position of this gene in this region (relative to the start of the region)<br></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">GENE_end</span></code> - the end position of this gene in this region (genome co-ordinates)<br></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">GENE_relative_end</span></code> - the end position of this gene in this region (relative to the start of the region)<br></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">GENE_strand</span></code> - the strand for this gene in this region<br></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">GENE_match</span></code> - the closest reference retrovirus to this gene in this region<br></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">GENE_group</span></code> - the group of the closest reference retrovirus to this gene in this region<br></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">GENE_genus</span></code> - the genus of the closest reference retrovirus to this gene in this region<br></p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">orig_name</span></code> - the name of the region in the input table<br></p></li>
</ul>
<p>If not all genes are screened for the table will not have the columns for this gene.</p>
<p>A bed file (<code class="docutils literal notranslate"><span class="pre">ERV_regions_final.bed</span></code>) is generated with the co-ordinates of the identified regions and a FASTA file (<code class="docutils literal notranslate"><span class="pre">ERV_regions_final.fasta</span></code>) containing their sequences.</p>
<div class="section" id="summariseervregions">
<h3>summariseERVRegions<a class="headerlink" href="#summariseervregions" title="Permalink to this headline">¶</a></h3>
<p><strong>Input Files</strong></p>
<p><strong>Output Files</strong></p>
<p><strong>Parameters</strong></p>
</div>
</div>
<div class="section" id="id1">
<h2>ERVRegions<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<p><strong>Input Files</strong></p>
<p><strong>Output Files</strong></p>
<p><strong>Parameters</strong></p>
</div>
<div class="section" id="full">
<h2>Full<a class="headerlink" href="#full" title="Permalink to this headline">¶</a></h2>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Katy Brown

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>