
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>What is CrossMap ? &#8212; CrossMap 0.5.2 documentation</title>
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="nav-item nav-item-0"><a href="#">CrossMap 0.5.2 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">What is CrossMap ?</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="toctree-wrapper compound">
</div>
<a class="reference internal image-reference" href="_images/logo.png"><img alt="alternate text" src="_images/logo.png" style="width: 375.0px; height: 75.0px;" /></a>
<div class="section" id="what-is-crossmap">
<h1>What is CrossMap ?<a class="headerlink" href="#what-is-crossmap" title="Permalink to this headline">¶</a></h1>
<p>CrossMap is a program for genome coordinates conversion between <em>different assemblies</em>
(such as <a class="reference external" href="http://www.ncbi.nlm.nih.gov/assembly/2928/">hg18 (NCBI36)</a>   &lt;=&gt; <a class="reference external" href="http://www.ncbi.nlm.nih.gov/assembly/2758/">hg19 (GRCh37)</a>).
It supports commonly used file formats including <a class="reference external" href="https://samtools.github.io/hts-specs/SAMv1.pdf">BAM</a>, <a class="reference external" href="https://en.wikipedia.org/wiki/CRAM_(file_format)">CRAM</a>, <a class="reference external" href="https://en.wikipedia.org/wiki/SAM_(file_format)">SAM</a>, <a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/wiggle.html">Wiggle</a>, <a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/bigWig.html">BigWig</a>, <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED</a>, <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format3">GFF</a>, <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format4">GTF</a>, <a class="reference external" href="https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/">MAF</a> <a class="reference external" href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">VCF</a>, and <a class="reference external" href="https://sites.google.com/site/gvcftools/home/about-gvcf">gVCF</a>.</p>
</div>
<div class="section" id="how-crossmap-works">
<h1>How CrossMap works?<a class="headerlink" href="#how-crossmap-works" title="Permalink to this headline">¶</a></h1>
<a class="reference internal image-reference" href="_images/howitworks.png"><img alt="alternate text" src="_images/howitworks.png" style="width: 510.0px; height: 212.5px;" /></a>
</div>
<div class="section" id="release-history">
<h1>Release history<a class="headerlink" href="#release-history" title="Permalink to this headline">¶</a></h1>
<p><strong>12/08/2020: Release version 0.5.2</strong></p>
<p>Add ‘–no-comp-alleles’ flag to <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">vcf</span></code> and <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">gvcf</span></code>. If set, CrossMap does not check if the “reference allele” is different from the “alternative allele”.</p>
<p><strong>08/19/2020: Release version 0.5.1</strong></p>
<p>In <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">region</span></code>: keep additional columns (columns after the 3rd column) of the original BED file after conversion.</p>
<p><strong>08/14/2020: Release version 0.5.0</strong></p>
<p>Add <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">region</span></code> function to convert large genomic regions. Unlike the <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">bed</span></code> function which splits big genomic regions, <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">region</span></code> tries to convert the big genomic region as a whole.</p>
<p><strong>07/09/2020: Release version 0.4.3</strong></p>
<p>Structural Variants VCF files often use INFO/END field to indicate the end of a deletion. v0.4.3 updates “END” coordinate in the INFO field.</p>
<p><strong>05/04/2020: Release version 0.4.2</strong></p>
<p>Support <a class="reference external" href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531812-GVCF-Genomic-Variant-Call-Format">GVCF</a> file conversion.</p>
<p><strong>03/24/2020: Release version 0.4.1</strong></p>
<p>Deal with consecutive TABs in the input MAF file.</p>
<p><strong>10/09/2019: Release version 0.3.8</strong></p>
<p>The University of California holds the copyrights in the UCSC chain files. As requested by UCSC, all UCSC generated chain files will be permanently removed from this website and the CrossMap distributions.</p>
<p><strong>07/22/2019: Release version 0.3.6</strong></p>
<ol class="arabic simple">
<li><p>Support MAF (mutation annotation format).</p></li>
<li><p>Fix error “TypeError: AlignmentHeader does not support item assignment (use header.to_dict()” when lifting over BAM files. User does not need to downgrade pysam to 0.13.0 to lift over BAM files.</p></li>
</ol>
<p><strong>04/01/2019: Release version 0.3.4</strong></p>
<p>Fix bugs when chromosome IDs (of the source genome) in chain file do not have ‘chr’ prefix (such as “GRCh37ToHg19.over.chain.gz”). This version also allows CrossMap to detect if a VCF mapping was inverted, and if so reverse complements the alternative allele (Thanks to Andrew Yates). Improve wording.</p>
<p><strong>01/07/2019: Release version 0.3.3</strong></p>
<p>Version 0.3.3 is exactly the same as Version 0.3.2. The reason to release this version is that CrossMap-0.3.2.tar.gz was broken when uploading to pypi.</p>
<p><strong>12/14/18: Release version 0.3.2</strong></p>
<p>Fix the key error problem (e.g  <em>KeyError: “sequence ‘b’7_KI270803v1_alt’’ not present”</em>). This error happens when a locus from the original assembly is mapped to an “alternative”, “unplaced” or “unlocalized” contig in the target assembly, and this “target contig” does not exist in your target_ref.fa. In version 0.3.2, such loci will be silently skipped and saved to the “.unmap” file.</p>
<p><strong>11/05/18: Release version 0.3.0</strong></p>
<ol class="arabic simple">
<li><p>v0.3.0 or newer will Support Python3. Previous versions support Python2.7.*</p></li>
<li><p>add <a class="reference external" href="https://github.com/deeptools/pyBigWig">pyBigWig</a> as a dependency.</p></li>
</ol>
</div>
<div class="section" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip3</span> <span class="n">install</span> <span class="n">CrossMap</span>  <span class="c1">#Install CrossMap supporting Python3</span>
<span class="n">pip3</span> <span class="n">install</span> <span class="n">CrossMap</span> <span class="o">--</span><span class="n">upgrade</span>        <span class="c1">#upgrade CrossMap supporting Python3</span>

<span class="n">pip2</span> <span class="n">install</span> <span class="n">CrossMap</span>  <span class="c1">#Install CrossMap supporting Python2.7.*</span>
<span class="n">pip2</span> <span class="n">install</span> <span class="n">CrossMap</span> <span class="o">--</span><span class="n">upgrade</span>        <span class="c1">#upgrade CrossMap supporting Python2.7.*</span>
</pre></div>
</div>
</div>
<div class="section" id="input-and-output">
<h1>Input and Output<a class="headerlink" href="#input-and-output" title="Permalink to this headline">¶</a></h1>
<div class="section" id="chain-file">
<h2>Chain file<a class="headerlink" href="#chain-file" title="Permalink to this headline">¶</a></h2>
<p>A <a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/chain.html">chain file</a> describes a pairwise alignment between two reference assemblies. UCSC and Ensemble chain files are available:</p>
<p><strong>UCSC chain files</strong></p>
<blockquote>
<div><ul class="simple">
<li><p>Chain files from hg38 (GRCh38) to hg19 and all other organisms: <a class="reference external" href="http://hgdownload.soe.ucsc.edu/goldenPath/hg38/liftOver/">http://hgdownload.soe.ucsc.edu/goldenPath/hg38/liftOver/</a></p></li>
<li><p>Chain File from hg19 (GRCh37) to hg17/hg18/hg38 and all other organisms: <a class="reference external" href="http://hgdownload.soe.ucsc.edu/goldenPath/hg19/liftOver/">http://hgdownload.soe.ucsc.edu/goldenPath/hg19/liftOver/</a></p></li>
<li><p>Chain File from mm10 (GRCm38) to mm9 and all other organisms: <a class="reference external" href="http://hgdownload.soe.ucsc.edu/goldenPath/mm10/liftOver/">http://hgdownload.soe.ucsc.edu/goldenPath/mm10/liftOver/</a></p></li>
</ul>
</div></blockquote>
<p><strong>Ensembl chain files</strong></p>
<blockquote>
<div><ul class="simple">
<li><p>Human to Human: <a class="reference external" href="ftp://ftp.ensembl.org/pub/assembly_mapping/homo_sapiens/">ftp://ftp.ensembl.org/pub/assembly_mapping/homo_sapiens/</a></p></li>
<li><p>Mouse to Mouse: <a class="reference external" href="ftp://ftp.ensembl.org/pub/assembly_mapping/mus_musculus/">ftp://ftp.ensembl.org/pub/assembly_mapping/mus_musculus/</a></p></li>
<li><p>Other organisms: <a class="reference external" href="ftp://ftp.ensembl.org/pub/assembly_mapping/">ftp://ftp.ensembl.org/pub/assembly_mapping/</a></p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="user-input-file">
<h2>User Input file<a class="headerlink" href="#user-input-file" title="Permalink to this headline">¶</a></h2>
<p>CrossMap supports the following file formats.</p>
<ol class="arabic simple">
<li><p><a class="reference external" href="http://samtools.sourceforge.net/SAMv1.pdf">BAM</a>, <a class="reference external" href="https://samtools.github.io/hts-specs/CRAMv3.pdf">CRAM</a>, or <a class="reference external" href="http://samtools.sourceforge.net/SAMv1.pdf/">SAM</a></p></li>
<li><p><a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED</a> or BED-like. (BED file must have at least ‘chrom’, ‘start’, ‘end’)</p></li>
<li><p><a class="reference external" href="http://genome.ucsc.edu/goldenPath/help/wiggle.html">Wiggle</a> (“variableStep”, “fixedStep” and “bedGraph” formats are supported)</p></li>
<li><p><a class="reference external" href="http://genome.ucsc.edu/goldenPath/help/bigWig.html">BigWig</a></p></li>
<li><p><a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format3">GFF</a> or <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format4">GTF</a></p></li>
<li><p><a class="reference external" href="http://vcftools.sourceforge.net/index.html">VCF</a></p></li>
<li><p><a class="reference external" href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531812-GVCF-Genomic-Variant-Call-Format">GVCF</a></p></li>
<li><p><a class="reference external" href="https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/">MAF</a></p></li>
</ol>
</div>
<div class="section" id="output-file">
<h2>Output file<a class="headerlink" href="#output-file" title="Permalink to this headline">¶</a></h2>
<p>The format of output files depends on the input</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 14%" />
<col style="width: 86%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Input_format</p></th>
<th class="head"><p>Output_format</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>BED</p></td>
<td><p>BED (Genome coordinates will be updated)</p></td>
</tr>
<tr class="row-odd"><td><p>BAM</p></td>
<td><p>BAM (Genome coordinates, header section, all SAM flags, insert size will be updated)</p></td>
</tr>
<tr class="row-even"><td><p>CRAM</p></td>
<td><p>BAM (require pysam &gt;= 0.8.2)</p></td>
</tr>
<tr class="row-odd"><td><p>SAM</p></td>
<td><p>SAM (Genome coordinates, header section, all SAM flags, insert size will be updated)</p></td>
</tr>
<tr class="row-even"><td><p>Wiggle</p></td>
<td><p>BigWig</p></td>
</tr>
<tr class="row-odd"><td><p>BigWig</p></td>
<td><p>BigWig</p></td>
</tr>
<tr class="row-even"><td><p>GFF</p></td>
<td><p>GFF (Genome coordinates will be updated to the target assembly)</p></td>
</tr>
<tr class="row-odd"><td><p>GTF</p></td>
<td><p>GTF (Genome coordinates will be updated to the target assembly)</p></td>
</tr>
<tr class="row-even"><td><p>VCF</p></td>
<td><p>VCF (header section, Genome coordinates, reference alleles will be updated)</p></td>
</tr>
<tr class="row-odd"><td><p>GVCF</p></td>
<td><p>GVCF (header section, Genome coordinates, reference alleles will be updated)</p></td>
</tr>
<tr class="row-even"><td><p>MAF</p></td>
<td><p>MAF (Genome coordinates and reference alleles will be updated)</p></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h1>
<p>Run CrossMap.py <strong>without</strong> any arguments will print a help message</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$CrossMap.py

Program: CrossMap (v0.5.2)

Description:
  CrossMap is a program to convert genome coordinates between different reference
  assemblies(e.g. from human hg19 to hg38 or vice versa). The supported file
  formats include BAM, BED, BigWig, CRAM, GFF, GTF, GVCF, MAF (mutation annotation
  format), SAM, Wiggle, and VCF.

Usage: CrossMap.py &lt;command&gt; [options]

        bam    convert BAM, CRAM or SAM format file.
        bed    convert BED, bedGraph or other BED-like files.
        bigwig convert BigWig file.
        gff    convert GFF or GTF format file.
        gvcf   convert GVCF file.
        maf    convert MAF (mutation annotation format) file.
        region convert big genomic regions (in BED format) such as CNV blocks.
        vcf    convert VCF file.
        wig    convert Wiggle or bedGraph format file.
</pre></div>
</div>
<div class="section" id="convert-bed-format-files">
<h2>Convert BED format files<a class="headerlink" href="#convert-bed-format-files" title="Permalink to this headline">¶</a></h2>
<p>A <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED</a> (Browser Extensible Data) file
is a tab-delimited text file describing genome regions or gene annotations.
It consists of one line per feature, each containing 3-12 columns.
CrossMap converts BED files with less than 12 columns to a different assembly by updating the
chromosome and genome coordinates only; all other columns remain unchanged. Regions from old
assembly mapping to multiple locations to the new assembly will be split.  For 12-columns BED
files, all columns will be updated accordingly except the 4th column (name of bed line), 5th
column (score value) and 9th column (RGB value describing the display color). 12-column BED
files usually define multiple blocks (e.g. exons); if any of the exons fails to map to a new
assembly, the whole BED line is skipped.</p>
<p>The input BED file can be plain text file, compressed file with extension of .gz, .Z, .z,
.bz, .bz2 and .bzip2, or even a URL pointing to accessible remote files (<a class="reference external" href="http://">http://</a>, <a class="reference external" href="https://">https://</a>
and <a class="reference external" href="ftp://">ftp://</a>). Compressed remote files are not supported. The output is a BED format file with
exact the same number of columns as the original one.</p>
<p>Standard <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED</a> format has 12 columns, but CrossMap also supports BED-like formats:</p>
<ul class="simple">
<li><p>BED3: The first 3 columns (“chrom”, “start”, “end”) of BED format file.</p></li>
<li><p>BED6: The first 6 columns (“chrom”, “start”, “end”, “name”, “score”, “strand”) of BED format file.</p></li>
<li><p>Other: Format has at least 3 columns (“chrom”, “start”, “end”) and no more than 12 columns. All other columns are arbitrary.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ol class="arabic simple">
<li><p>For BED-like formats mentioned above, CrossMap only updates the “chrom”, “start”, “end”, and “strand” columns. All other columns will be kept AS-IS.</p></li>
<li><p>Lines starting with ‘#’, ‘browser’, ‘track’ will be skipped.</p></li>
<li><p>Lines less than 3 columns will be skipped.</p></li>
<li><p>The 2nd and 3rd columns must be integers.</p></li>
<li><p>The “+” strand is assumed if no strand information is found.</p></li>
<li><p>For standard BED format (12 columns). If any of the defined exon blocks cannot be uniquely mapped to target assembly, the whole entry will be skipped.</p></li>
<li><p>The “input_chain_file” and “input_bed_file” can be regular or compressed (.gz, .Z, .z, .bz, .bz2, .bzip2) file, local file or URL (<a class="reference external" href="http://">http://</a>, <a class="reference external" href="https://">https://</a>, <a class="reference external" href="ftp://">ftp://</a>) pointing to remote file.</p></li>
<li><p>If the output_file is not specified, results will be printed to screen (console). In this case, the original bed entries (include entries failed to convert) were also printed out.</p></li>
<li><p>If the input region cannot be consecutively mapped to the target assembly, it will be split.</p></li>
<li><p>The *.unmap file contains regions that cannot be unambiguously converted.</p></li>
</ol>
</div>
<p><strong>Example 1</strong></p>
<p>run <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">bed</span></code> with <strong>no</strong> <em>output_file</em>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap.py bed hg18ToHg19.over.chain.gz test.hg18.bed3

# Conversion results were printed to screen directly (column1-3 are hg18 based, column5-7 are hg19 based)::
chr1   142614848       142617697       -&gt;      chr1    143903503       143906352
chr1   142617697       142623312       -&gt;      chr1    143906355       143911970
chr1   142623313       142623350       -&gt;      chr1    143911971       143912008
</pre></div>
</div>
<p><strong>Example 2</strong></p>
<p>run <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">bed</span></code> with <em>output_file</em> (test.hg19.bed3) specified:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap.py bed hg18ToHg19.over.chain.gz test.hg18.bed3 test.hg19.bed3

$ cat test.hg19.bed3
chr1   143903503       143906352
chr1   143906355       143911970
chr1   143911971       143912008
</pre></div>
</div>
<p><strong>Example 3</strong></p>
<p>One input region was split because it cannot be mapped consecutively to the target assembly:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap.py bed hg18ToHg19.over.chain.gz test.hg18.bed3

chr10  81369946        81370453        +       -&gt;      chr10   81380000        81380507        +
chr10  81370483        81371363        +       -&gt;      chr10   81380539        81381419        +
chr10  81371363        81371365        +       -&gt;      chr10   62961832        62961834        +
chr10  81371412        81371432        +       (split.1:chr10:81371412:81371422:+)     chr10   62961775        62961785        +
chr10  81371412        81371432        +       (split.2:chr10:81371422:81371432:+)     chrX    63278348        63278358        +
</pre></div>
</div>
<p><strong>Example 4</strong></p>
<p><a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/bedgraph.html">BedGraph</a> format file can be converted using either <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">bed</span></code> or <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">wig</span></code>,
however, the output formats are different:</p>
<ul class="simple">
<li><p>Use <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">bed</span></code> command to convert a bedGraph file, output is a <strong>bedGraph</strong> file.</p></li>
<li><p>Use <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">wig</span></code> command to convert a bedGraph file, output is a <strong>bigWig</strong> file.</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap.py bed hg19ToHg38.over.chain.gz 4_hg19.bgr

chrX   5873316 5873391 2.0     -&gt;      chrX    5955275 5955350 2.0
chrX   5873673 5873710 0.8     -&gt;      chrX    5955632 5955669 0.8
chrX   5873710 5873785 1.4     -&gt;      chrX    5955669 5955744 1.4
chrX   5873896 5873929 0.9     -&gt;      chrX    5955855 5955888 0.9
chrX   5873929 5874004 1.5     -&gt;      chrX    5955888 5955963 1.5
chrX   5874230 5874471 0.3     -&gt;      chrX    5956189 5956430 0.3
chrX   5874471 5874518 0.9     -&gt;      chrX    5956430 5956477 0.9

$ python3 CrossMap.py wig hg19ToHg38.over.chain.gz 4_hg19.bgr output_hg38
@ 2018-11-06 00:09:11: Read chain_file:  hg19ToHg38.over.chain.gz
@ 2018-11-06 00:09:12: Liftover wiggle file: 4_hg19.bgr ==&gt; output_hg38.bgr
@ 2018-11-06 00:09:12: Merging overlapped entries in bedGraph file ...
@ 2018-11-06 00:09:12: Sorting bedGraph file:output_hg38.bgr
@ 2018-11-06 00:09:12: Writing header to &quot;output_hg38.bw&quot; ...
@ 2018-11-06 00:09:12: Writing entries to &quot;output_hg38.bw&quot; ...
</pre></div>
</div>
<p><strong>Example 5</strong></p>
<p>Use <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">region</span></code> command to convert large genomic regions (such as <a class="reference external" href="https://en.wikipedia.org/wiki/Copy_number_variation">CNV</a> blocks) in BED format.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># a genomic region of 3.48Mb
$ cat test.bed
chr2   239716679       243199373
</pre></div>
</div>
<p>If we use <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">bed</span></code> command to convert this 3.48 Mb region. It will be split into 74 small blocks:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$CrossMap.py bed GRCh37_to_GRCh38.chain.gz  test.bed

chr2   239716679       243199373       (split.1:chr2:239716679:239801978:+)    chr2    238808038       238893337
chr2   239716679       243199373       (split.2:chr2:239831978:240205681:+)    chr2    238910282       239283985
chr2   239716679       243199373       (split.3:chr2:240205681:240319336:+)    chr2    239283986       239397641
... (split 74 times)
</pre></div>
</div>
<p>If we use <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">region</span></code> command to convert this 3.48Mb region. Note: “-r” (the minimum ratio of bases that must remap) is 0.85 by default:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$CrossMap.py region GRCh37_to_GRCh38.chain.gz  test.bed

chr2   239716679       243199373       -&gt;      chr2    238808038       242183529       map_ratio=0.9622
</pre></div>
</div>
<p>If we increase -r to 0.99, this region will fail:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$CrossMap.py region GRCh37_to_GRCh38.chain.gz  test.bed -r 0.99

chr2   239716679       243199373       Fail    map_ratio=0.9622
</pre></div>
</div>
</div>
<div class="section" id="convert-bam-cram-sam-format-files">
<span id="bam-conversion"></span><h2>Convert BAM/CRAM/SAM format files<a class="headerlink" href="#convert-bam-cram-sam-format-files" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml#5">SAM</a> (Sequence Alignment Map) format
is a generic format for storing sequencing alignments, and BAM is the binary and compressed
version of SAM (<a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/25/16/2078.full">Li et al., 2009</a>).
<a class="reference external" href="https://samtools.github.io/hts-specs/CRAMv3.pdf">CRAM</a> was designed to be an efficient reference-based
alternative to the <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml#5">SAM</a> and BAM file formats
Most high-throughput sequencing  (HTS) alignments were in SAM/BAM format and many HTS analysis
tools work with SAM/BAM format. CrossMap updates chromosomes, genome coordinates, header
sections, and all SAM flags accordingly.  CrossMap’s version number is inserted into
the header section, along with the names of the original BAM file and the chain file.  For
pair-end sequencing, insert size is also recalculated. The input BAM file should be sorted
and indexed properly using Samtools (<a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/25/16/2078.full">Li et al., 2009</a>).
The output format is determined by the input format and BAM output will be sorted and indexed automatically.</p>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">bam</span></code> without any arguments will print help a message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Usage</span><span class="p">:</span>
<span class="n">CrossMap</span><span class="o">.</span><span class="n">py</span> <span class="n">bam</span>  <span class="o">&lt;</span><span class="n">chain_file</span><span class="o">&gt;</span>  <span class="o">&lt;</span><span class="nb">input</span><span class="o">.</span><span class="n">bam</span><span class="o">&gt;</span> <span class="p">[</span><span class="n">output_file</span><span class="p">]</span> <span class="p">[</span><span class="n">options</span><span class="p">]</span>

<span class="o">..</span> <span class="n">NOTE</span><span class="p">::</span>
   <span class="n">If</span> <span class="n">output_file</span> <span class="ow">is</span> <span class="s1">&#39;STDOUT&#39;</span><span class="p">,</span><span class="s1">&#39;-&#39;</span> <span class="ow">or</span> <span class="n">missing</span><span class="p">,</span> <span class="n">CrossMap</span> <span class="n">will</span> <span class="n">write</span> <span class="n">the</span> <span class="n">BAM</span> <span class="n">file</span> <span class="n">to</span> <span class="n">Stdout</span>

<span class="n">Options</span><span class="p">:</span>
  <span class="o">-</span><span class="n">m</span> <span class="n">INSERT_SIZE</span><span class="p">,</span> <span class="o">--</span><span class="n">mean</span><span class="o">=</span><span class="n">INSERT_SIZE</span>
                        <span class="n">Average</span> <span class="n">insert</span> <span class="n">size</span> <span class="n">of</span> <span class="n">pair</span><span class="o">-</span><span class="n">end</span> <span class="n">sequencing</span> <span class="p">(</span><span class="n">bp</span><span class="p">)</span><span class="o">.</span>
                        <span class="p">{</span><span class="n">default</span><span class="o">=</span><span class="mf">200.0</span><span class="p">}</span>
  <span class="o">-</span><span class="n">s</span> <span class="n">INSERT_SIZE_STDEV</span><span class="p">,</span> <span class="o">--</span><span class="n">stdev</span><span class="o">=</span><span class="n">INSERT_SIZE_STDEV</span>
                       <span class="n">Stanadard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">insert</span> <span class="n">size</span><span class="o">.</span> <span class="p">{</span><span class="n">default</span><span class="o">=</span><span class="mf">30.0</span><span class="p">}</span>
  <span class="o">-</span><span class="n">t</span> <span class="n">INSERT_SIZE_FOLD</span><span class="p">,</span> <span class="o">--</span><span class="n">times</span><span class="o">=</span><span class="n">INSERT_SIZE_FOLD</span>
                        <span class="n">A</span> <span class="n">mapped</span> <span class="n">pair</span> <span class="ow">is</span> <span class="n">considered</span> <span class="k">as</span> <span class="s2">&quot;proper pair&quot;</span> <span class="k">if</span> <span class="n">both</span>
                        <span class="n">ends</span> <span class="n">mapped</span> <span class="n">to</span> <span class="n">different</span> <span class="n">strand</span> <span class="ow">and</span>
                        <span class="n">the</span> <span class="n">distance</span> <span class="n">between</span> <span class="n">them</span> <span class="ow">is</span> <span class="n">less</span> <span class="n">than</span> <span class="s1">&#39;-t&#39;</span> <span class="o">*</span> <span class="n">stdev</span>
                        <span class="kn">from</span> <span class="nn">the</span> <span class="n">mean</span><span class="o">.</span> <span class="p">{</span><span class="n">default</span><span class="o">=</span><span class="mf">3.0</span><span class="p">}</span>
  <span class="o">-</span><span class="n">a</span><span class="p">,</span> <span class="o">--</span><span class="n">append</span><span class="o">-</span><span class="n">tags</span>     <span class="n">Add</span> <span class="n">tag</span> <span class="n">to</span> <span class="n">each</span> <span class="n">alignment</span><span class="o">.</span>
</pre></div>
</div>
<p><strong>Example</strong></p>
<p>Convert BAM from hg19 to hg18:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># add optional tags using &#39;-a&#39; (recommend always use &#39;-a&#39; option)

$ CrossMap.py bam -a ../data/hg19ToHg18.over.chain.gz test.hg19.bam test.hg18
Insert size = 200.000000
Insert size stdev = 30.000000
Number of stdev from the mean = 3.000000
Add tags to each alignment = True
@ 2016-10-07 15:29:06: Read chain_file:  ../data/hg19ToHg18.over.chain.gz
@ 2016-10-07 15:29:07: Liftover BAM file: test.hg19.bam ==&gt; test.hg18.bam
@ 2016-10-07 15:29:14: Done!
@ 2016-10-07 15:29:14: Sort &quot;test.hg18.bam&quot; ...
@ 2016-10-07 15:29:15: Index &quot;test.hg18.sorted.bam&quot; ...
Total alignments:99914
       QC failed: 0
       R1 unique, R2 unique (UU): 96094
       R1 unique, R2 unmapp (UN): 3579
       R1 unique, R2 multiple (UM): 0
       R1 multiple, R2 multiple (MM): 0
       R1 multiple, R2 unique (MU): 233
       R1 multiple, R2 unmapped (MN): 8
       R1 unmap, R2 unmap (NN): 0
       R1 unmap, R2 unique (NU): 0
       R1 unmap, R2 multiple (NM): 0
</pre></div>
</div>
<p># BAM/SAM header sections was updated:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools view -H  test.hg19.bam
@SQ    SN:chr1 LN:249250621
@SQ    SN:chr2 LN:243199373
@SQ    SN:chr3 LN:198022430
...
@SQ    SN:chrX LN:155270560
@SQ    SN:chrY LN:59373566
@SQ    SN:chrM LN:16571
@RG    ID:Sample_618545BE      SM:Sample_618545BE      LB:Sample_618545BE      PL:Illumina
@PG    ID:bwa  PN:bwa  VN:0.6.2-r126

$ samtools view -H  test.hg18.bam
@HD    VN:1.0  SO:coordinate
@SQ    SN:chr1 LN:247249719
@SQ    SN:chr10        LN:135374737
@SQ    SN:chr11        LN:134452384
...
@SQ    SN:chrX LN:154913754
@SQ    SN:chrX_random  LN:1719168
@SQ    SN:chrY LN:57772954
@RG    ID:Sample_618545BE      SM:Sample_618545BE      LB:Sample_618545BE      PL:Illumina
@PG    PN:bwa  ID:bwa  VN:0.6.2-r126
@PG    ID:CrossMap     VN:0.5.0
@CO    Liftover from original BAM/SAM file: test.hg19.bam
@CO    Liftover is based on the chain file: ../test/hg19ToHg18.over.chain.gz
</pre></div>
</div>
<p><strong>Optional tags:</strong></p>
<dl class="simple">
<dt>Q</dt><dd><p>QC. QC failed.</p>
</dd>
<dt>N</dt><dd><p>Unmapped. Originally unmapped or originally mapped but failed to lift over to new assembly.</p>
</dd>
<dt>M</dt><dd><p>Multiple mapped. Alignment can be lifted over to multiple places.</p>
</dd>
<dt>U</dt><dd><p>Unique mapped. Alignment can be lifted over to only 1 place.</p>
</dd>
</dl>
<p><strong>Tags for pair-end sequencing include:</strong></p>
<ul class="simple">
<li><p>QF = QC failed</p></li>
<li><p>NN = both read1 and read2 unmapped</p></li>
<li><p>NU = read1 unmapped, read2 unique mapped</p></li>
<li><p>NM = read1 unmapped, multiple mapped</p></li>
<li><p>UN = read1 uniquely mapped, read2 unmap</p></li>
<li><p>UU = both read1 and read2 uniquely mapped</p></li>
<li><p>UM = read1 uniquely mapped, read2 multiple mapped</p></li>
<li><p>MN = read1 multiple mapped, read2 unmapped</p></li>
<li><p>MU = read1 multiple mapped, read2 unique mapped</p></li>
<li><p>MM = both read1 and read2 multiple mapped</p></li>
</ul>
<p><strong>Tags for single-end sequencing include:</strong></p>
<ul class="simple">
<li><p>QF = QC failed</p></li>
<li><p>SN = unmaped</p></li>
<li><p>SM = multiple mapped</p></li>
<li><p>SU = uniquely mapped</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ol class="arabic simple">
<li><p>All alignments (mapped, partial mapped, unmapped, QC failed) will write to one file. Users can filter them by tags.</p></li>
<li><p>The header section will be updated to the target assembly.</p></li>
<li><p>Genome coordinates and all SAM flags in the alignment section will be updated to the target assembly.</p></li>
<li><p>If the input is a CRAM file, pysam version should &gt;= 0.8.2</p></li>
<li><p>Optional fields in the alignment section will not update.</p></li>
</ol>
</div>
</div>
<div class="section" id="convert-wiggle-format-files">
<h2>Convert Wiggle format files<a class="headerlink" href="#convert-wiggle-format-files" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="http://genome.ucsc.edu/goldenPath/help/wiggle.html">Wiggle</a> (WIG) format is useful for
displaying continuous data such as GC content and the reads intensity of high-throughput sequencing data.
BigWig is a self-indexed binary-format Wiggle file and has the advantage of supporting random access.
Input wiggle data can be in variableStep (for data with irregular intervals) or fixedStep
(for data with regular intervals). Regardless of the input, the output files are always in bedGraph
format. We export files in bedGraph format because it’s more compact than wiggle format, and more importantly,
CrossMap internally transforms wiggle into bedGraph to increase running speed.</p>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">wig</span></code> without any arguments will print a help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap.py  wig

Usage
-----
  CrossMap.py  wig  &lt;chain_file&gt;  &lt;input.wig&gt;  &lt;output_prefix&gt;

Description
-----------
  Convert wiggle format file. The &quot;chain_file&quot; can be regular or compressed (*.gz,
  *.Z, *.z, *.bz, *.bz2, *.bzip2) file, local file or URL (http://, https://,
  ftp://) pointing to remote file.  Both &quot;variableStep&quot; and &quot;fixedStep&quot; wiggle
  lines are supported. Wiggle format:
  http://genome.ucsc.edu/goldenPath/help/wiggle.html

Example
-------
  CrossMap.py wig hg18ToHg19.over.chain.gz test.hg18.wig test.hg19
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To improve performance, this script calls <a class="reference external" href="http://www.gnu.org/software/coreutils/manual/html_node/sort-invocation.html">GNU “sort”</a> command internally. If the “sort” command is not callable, CrossMap will exit.</p>
</div>
</div>
<div class="section" id="convert-bigwig-format-files">
<h2>Convert BigWig format files<a class="headerlink" href="#convert-bigwig-format-files" title="Permalink to this headline">¶</a></h2>
<p>If an input file is in BigWig format, the output is BigWig format if UCSC’s
‘<a class="reference external" href="http://hgdownload.cse.ucsc.edu/admin/exe/">wigToBigWig</a>’ executable can be found;
otherwise, the output file will be in bedGraph format.</p>
<p>After v0.3.0, UCSC’s <code class="code docutils literal notranslate"><span class="pre">wigToBigWig</span></code> command is no longer needed.</p>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">bigwig</span></code> without any arguments will print a help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap.py bigwig

Usage
-----
  CrossMap.py  bigwig  &lt;chain_file&gt;  &lt;input.bigwig&gt;  &lt;output_prefix&gt;

Description
-----------
  Convert BigWig format file. The &quot;chain_file&quot; can be regular or compressed (*.gz,
  *.Z, *.z, *.bz, *.bz2, *.bzip2) file, local file or URL (http://, https://,
  ftp://) pointing to remote file. Bigwig format:
  http://genome.ucsc.edu/goldenPath/help/bigWig.html

Example
-------
  CrossMap.py bigwig hg18ToHg19.over.chain.gz test.hg18.bw test.hg19
</pre></div>
</div>
<p>Example (Convert BigWig file from hg18 to hg19):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python CrossMap.py bigwig  hg19ToHg18.over.chain.gz  test.hg19.bw test.hg18
@ 2013-11-17 22:12:42: Read chain_file:  ../data/hg19ToHg18.over.chain.gz
@ 2013-11-17 22:12:44: Liftover bigwig file: test.hg19.bw ==&gt; test.hg18.bgr
@ 2013-11-17 22:15:38: Merging overlapped entries in bedGraph file ...
@ 2013-11-17 22:15:38: Sorting bedGraph file:test.hg18.bgr
@ 2013-11-17 22:15:39: Convert wiggle to bigwig ...
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To improve performance, this script calls <a class="reference external" href="http://www.gnu.org/software/coreutils/manual/html_node/sort-invocation.html">GNU “sort”</a> command internally. If “sort” command does not exist, CrossMap will exit.</p>
</div>
</div>
<div class="section" id="convert-gff-gtf-format-files">
<h2>Convert GFF/GTF format files<a class="headerlink" href="#convert-gff-gtf-format-files" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format3">GFF</a> (General Feature Format) is another
plain text file used to describe gene structure. <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format4">GTF</a>
(Gene Transfer Format) is a refined version of GTF. The first eight fields are the same as
GFF. Plain text, compressed plain text, and URLs pointing to remote files are all supported.
Only chromosome and genome coordinates are updated. The format of the output is determined from
the input.</p>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">gff</span></code> without any arguments will print a help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap.py  gff

Usage
-----
  CrossMap.py  gff  &lt;chain_file&gt;  &lt;input.gff&gt;  &lt;output_file&gt;

Description
-----------
  Convert GFF or GTF format file. The&quot;chain_file&quot; can be regular or compressed
  (*.gz, *.Z, *.z, *.bz, *.bz2, *.bzip2) file, local file or URL (http://,
  https://, ftp://) pointing to remote file. Input file must be in GFF or GTF
  format. GFF format: http://genome.ucsc.edu/FAQ/FAQformat.html#format3 GTF
  format: http://genome.ucsc.edu/FAQ/FAQformat.html#format4

Example1 (write output to file)
-------------------------------
  CrossMap.py gff  hg19ToHg18.over.chain.gz test.hg19.gtf test.hg18.gtf

Example2 (write output to the screen)
---------------------------------
  CrossMap.py gff hg19ToHg18.over.chain.gz test.hg19.gtf
</pre></div>
</div>
<p>Example (Convert GTF file from hg19 to hg18):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python CrossMap.py gff  hg19ToHg18.over.chain.gz test.hg19.gtf test.hg18.gtf
@ 2013-11-17 20:44:47: Read chain_file:  ../data/hg19ToHg18.over.chain.gz

$ head test.hg19.gtf
chr1   hg19_refGene    CDS     48267145        48267291        0.000000        -       0       gene_id &quot;NM_001194986&quot;; transcript_id &quot;NM_001194986&quot;;
chr1   hg19_refGene    exon    66081691        66081907        0.000000        +       .       gene_id &quot;NM_002303&quot;; transcript_id &quot;NM_002303&quot;;
chr1   hg19_refGene    CDS     145334684       145334792       0.000000        +       2       gene_id &quot;NM_001039703&quot;; transcript_id &quot;NM_001039703&quot;;
chr1   hg19_refGene    exon    172017752       172017890       0.000000        +       .       gene_id &quot;NM_001136127&quot;; transcript_id &quot;NM_001136127&quot;;
chr1   hg19_refGene    CDS     206589249       206589333       0.000000        +       2       gene_id &quot;NM_001170637&quot;; transcript_id &quot;NM_001170637&quot;;
chr1   hg19_refGene    exon    210573812       210574006       0.000000        +       .       gene_id &quot;NM_001170580&quot;; transcript_id &quot;NM_001170580&quot;;
chr1   hg19_refGene    CDS     235850249       235850347       0.000000        -       0       gene_id &quot;NM_000081&quot;; transcript_id &quot;NM_000081&quot;;
chr1   hg19_refGene    CDS     235880012       235880078       0.000000        -       1       gene_id &quot;NM_000081&quot;; transcript_id &quot;NM_000081&quot;;
chr1   hg19_refGene    exon    3417741 3417872 0.000000        -       .       gene_id &quot;NM_001409&quot;; transcript_id &quot;NM_001409&quot;;
chr1   hg19_refGene    exon    10190773        10190871        0.000000        +       .       gene_id &quot;NM_006048&quot;; transcript_id &quot;NM_006048&quot;;

$ head test.hg18.gtf
chr1   hg19_refGene    CDS     48039732        48039878        0.000000        -       0       gene_id &quot;NM_001194986&quot;; transcript_id &quot;NM_001194986&quot;;
chr1   hg19_refGene    exon    65854279        65854495        0.000000        +       .       gene_id &quot;NM_002303&quot;; transcript_id &quot;NM_002303&quot;;
chr1   hg19_refGene    CDS     144046041       144046149       0.000000        +       2       gene_id &quot;NM_001039703&quot;; transcript_id &quot;NM_001039703&quot;;
chr1   hg19_refGene    exon    170284375       170284513       0.000000        +       .       gene_id &quot;NM_001136127&quot;; transcript_id &quot;NM_001136127&quot;;
chr1   hg19_refGene    CDS     204655872       204655956       0.000000        +       2       gene_id &quot;NM_001170637&quot;; transcript_id &quot;NM_001170637&quot;;
chr1   hg19_refGene    exon    208640435       208640629       0.000000        +       .       gene_id &quot;NM_001170580&quot;; transcript_id &quot;NM_001170580&quot;;
chr1   hg19_refGene    CDS     233916872       233916970       0.000000        -       0       gene_id &quot;NM_000081&quot;; transcript_id &quot;NM_000081&quot;;
chr1   hg19_refGene    CDS     233946635       233946701       0.000000        -       1       gene_id &quot;NM_000081&quot;; transcript_id &quot;NM_000081&quot;;
chr1   hg19_refGene    exon    3407601 3407732 0.000000        -       .       gene_id &quot;NM_001409&quot;; transcript_id &quot;NM_001409&quot;;
chr1   hg19_refGene    exon    10113360        10113458        0.000000        +       .       gene_id &quot;NM_006048&quot;; transcript_id &quot;NM_006048&quot;;
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ol class="arabic simple">
<li><p>Each feature  (exon, intron, UTR, etc) is processed separately and independently, and we do NOT check if features originally belonging to the same gene were converted into the same gene.</p></li>
<li><p>If a user wants to lift over gene annotation files, use BED12 format.</p></li>
<li><p>If no output file was specified, the output will be printed to screen (console). In this case, items that failed to convert are also printed out.</p></li>
</ol>
</div>
</div>
<div class="section" id="convert-vcf-format-files">
<h2>Convert VCF format files<a class="headerlink" href="#convert-vcf-format-files" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41">VCF</a>
(variant call format) is a flexible and extendable line-oriented text format developed by
the <a class="reference external" href="http://www.1000genomes.org/">1000 Genome Project</a>. It is useful for representing single
nucleotide variants, indels, copy number variants, and structural variants. Chromosomes,
coordinates, and reference alleles are updated to a new assembly, and all the other fields
are not changed.</p>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">vcf</span></code> without any arguments will print a help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap.py  vcf

Usage:
CrossMap.py vcf &lt;chain_file&gt;  &lt;input.vcf&gt;  &lt;refGenome.fa&gt;  &lt;output_file&gt; [options]

Options:
  --no-comp-alleles  If set, CrossMap does NOT check if the reference allele is
                     different from the alternate allele.
 --compress          If set, compress the output VCF file by calling the
                     system &quot;gzip&quot;.
</pre></div>
</div>
<p>Example: filter out variants [reference_allele == alternative_allele]:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap.py  vcf  GRCh37_to_GRCh38.chain.gz  test02_hg19.vcf  hg38.fa  out.hg38.vcf
@ 2020-12-08 22:33:16: Read the chain file:  ../data/human/GRCh37_to_GRCh38.chain.gz
@ 2020-12-08 22:33:17: Filter out variants [reference_allele == alternative_allele] ...
@ 2020-12-08 22:33:17: Updating contig field ...
@ 2020-12-08 22:33:17: Lifting over ...
@ 2020-12-08 22:33:17: Total entries: 882
@ 2020-12-08 22:33:17: Failed to map: 2
</pre></div>
</div>
<p>Example: Keep variants [reference_allele == alternative_allele]. Turn on <code class="code docutils literal notranslate"><span class="pre">--no-comp-allele</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap.py  vcf  GRCh37_to_GRCh38.chain.gz  test02_hg19.vcf  hg38.fa  out.hg38.vcf --no-comp-allele
@ 2020-12-08 22:36:51: Read the chain file:  ../data/human/GRCh37_to_GRCh38.chain.gz
@ 2020-12-08 22:36:51: Keep variants [reference_allele == alternative_allele] ...
@ 2020-12-08 22:36:51: Updating contig field ...
@ 2020-12-08 22:36:51: Lifting over ...
@ 2020-12-08 22:36:51: Total entries: 882
@ 2020-12-08 22:36:51: Failed to map: 1
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ol class="arabic simple">
<li><p>Genome coordinates and reference alleles will be updated to target assembly.</p></li>
<li><p>Reference genome is the genome sequences of target assembly.</p></li>
<li><p>If the reference genome sequence file (../database/genome/hg18.fa) was not indexed, CrossMap will automatically index it (only the first time you run CrossMap).</p></li>
<li><p>Output files: <em>output_file</em> and <em>output_file.unmap</em>.</p></li>
<li><p>In the output VCF file, whether the chromosome IDs contain “chr” or not depends on the format of the input VCF file.</p></li>
</ol>
</div>
<p><strong>Interpretation of Failed tags</strong>:</p>
<ul class="simple">
<li><p>Fail(Multiple_hits) : This genomic location was mapped to two or more locations to the target assembly.</p></li>
<li><p>Fail(REF==ALT) : After liftover, the reference allele is the same as the alternative allele (i.e. this is NOT an SNP/variant after liftover). In version 0.5.2, this checking can be turned off by setting ‘–no-comp-alleles’.</p></li>
<li><p>Fail(Unmap) : Unable to map this location to the target assembly.</p></li>
<li><p>Fail(KeyError) : Unable to find the contig ID (or chromosome ID) from the reference genome sequence (of the target assembly).</p></li>
</ul>
</div>
<div class="section" id="convert-maf-format-files">
<h2>Convert MAF format files<a class="headerlink" href="#convert-maf-format-files" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://docs.gdc.cancer.gov/Encyclopedia/pages/Mutation_Annotation_Format/">MAF</a>
(mutation annotation format) files are tab-delimited
files that contain somatic and/or germline mutation annotations. Please do not confuse
with the <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format5">Multiple Alignment Format</a>.</p>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">maf</span></code> without any arguments will print a help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap.py  maf

usage
-----
  CrossMap.py maf  &lt;chain_file&gt;  &lt;input.maf&gt;  &lt;refGenome.fa&gt;  &lt;build_name&gt;
  &lt;output_file&gt;

Description
-----------
  Convert MAF format file. The &quot;chain_file&quot; and &quot;input.maf&quot; can be regular or
  compressed (*.gz, *.Z, *.z, *.bz, *.bz2, *.bzip2) file, local file or URL
  (http://, https://, ftp://) pointing to remote file. &quot;refGenome.fa&quot; is genome
  sequence file of *target assembly*. &quot;build_name&quot; is the name of the
  *target_assembly* (eg &quot;GRCh38&quot;)

Example
-------
 CrossMap.py  maf       hg19ToHg38.over.chain.gz  test.hg19.maf  hg38.fa  GRCh38 test.hg38.maf
</pre></div>
</div>
</div>
<div class="section" id="convert-gvcf-format-files">
<h2>Convert GVCF format files<a class="headerlink" href="#convert-gvcf-format-files" title="Permalink to this headline">¶</a></h2>
<p>GVCF file format is described in <a class="reference external" href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531812-GVCF-Genomic-Variant-Call-Format">here</a>.</p>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">gvcf</span></code> without any arguments will print a help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap.py  gvcf

Usage:
CrossMap.py gvcf &lt;chain_file&gt;  &lt;input.gvcf&gt;  &lt;refGenome.fa&gt;  &lt;output_file&gt; [options]

 Options:
  --no-comp-alleles  If set, CrossMap does NOT check if the reference allele is
                     different from the alternate allele.
 --compress          If set, compress the output gVCF file by calling the
                     system &quot;gzip&quot;.
</pre></div>
</div>
<p>Example (Convert GVCF file from hg19 to hg38):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap.py  gvcf  GRCh37_to_GRCh38.chain.gz  test10_hg19.gvcf   hg38.fa  out.hg38.gvcf
@ 2020-12-08 22:19:44: Read the chain file:  ../data/human/GRCh37_to_GRCh38.chain.gz
@ 2020-12-08 22:19:44: Filter out variants [reference_allele == alternative_allele] ...
@ 2020-12-08 22:19:44: Updating contig field ...
@ 2020-12-08 22:19:44: Lifting over ...
@ 2020-12-08 22:19:44: Total variants: 10
@ 2020-12-08 22:19:44: Variants failed to map: 1
@ 2020-12-08 22:19:44: Total non-variant regions: 22
@ 2020-12-08 22:19:44: Non-variant regions failed to map: 0
</pre></div>
</div>
</div>
<div class="section" id="convert-large-genomic-regions">
<h2>Convert large genomic regions<a class="headerlink" href="#convert-large-genomic-regions" title="Permalink to this headline">¶</a></h2>
<p>For <strong>large genomic regions</strong> such as CNV blocks, the <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">bed</span></code> will split each large region into smaller blocks that are 100% matched to the target assembly.
<code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">region</span></code> will NOT split large regions, instead, it will calculate the <strong>map ratio</strong> (i.e. {bases mapped to target genome} / {total bases in query region}). If the
<strong>map ratio</strong> is larger than the threshold specified by <code class="code docutils literal notranslate"><span class="pre">-r</span></code>, the coordinates will be converted to the target genome, otherwise, fails.</p>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">region</span></code> without any arguments will print a help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Usage</span>
<span class="o">-----</span>
<span class="n">CrossMap</span><span class="o">.</span><span class="n">py</span> <span class="n">region</span> <span class="o">&lt;</span><span class="n">chain_file</span><span class="o">&gt;</span>  <span class="o">&lt;</span><span class="n">regions</span><span class="o">.</span><span class="n">bed</span><span class="o">&gt;</span> <span class="p">[</span><span class="n">output_file</span><span class="p">]</span> <span class="p">[</span><span class="n">options</span><span class="p">]</span>

<span class="n">Examples</span><span class="p">:</span>
<span class="n">CrossMap</span><span class="o">.</span><span class="n">py</span> <span class="n">region</span> <span class="n">hg18ToHg19</span><span class="o">.</span><span class="n">over</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">gz</span> <span class="n">CNV</span><span class="o">.</span><span class="n">hg18</span><span class="o">.</span><span class="n">bed</span> <span class="n">CNV</span><span class="o">.</span><span class="n">hg19</span><span class="o">.</span><span class="n">bed</span>   <span class="c1"># write to file</span>
<span class="n">CrossMap</span><span class="o">.</span><span class="n">py</span> <span class="n">region</span> <span class="n">hg18ToHg19</span><span class="o">.</span><span class="n">over</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">gz</span> <span class="n">CNV</span><span class="o">.</span><span class="n">hg18</span><span class="o">.</span><span class="n">bed</span>   <span class="c1"># write to screen</span>

<span class="n">Options</span><span class="p">:</span>
  <span class="o">-</span><span class="n">r</span> <span class="n">MIN_MAP_RATIO</span><span class="p">,</span> <span class="o">--</span><span class="n">ratio</span><span class="o">=</span><span class="n">MIN_MAP_RATIO</span>
                        <span class="n">Minimum</span> <span class="n">ratio</span> <span class="n">of</span> <span class="n">bases</span> <span class="n">that</span> <span class="n">must</span> <span class="n">remap</span><span class="o">.</span> <span class="p">{</span><span class="n">default</span><span class="o">=</span><span class="mf">0.85</span><span class="p">}</span>
</pre></div>
</div>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$CrossMap.py region  GRCh37_to_GRCh38.chain.gz test11_hg19_region.bed

@ 2020-08-14 16:46:04: Read the chain file:  ../data/human/GRCh37_to_GRCh38.chain.gz
chr1   0       2500000 -&gt;      chr1    10000   2568561 map_ratio=0.9360
chr1   145394955       145807817       -&gt;      chr1    145627235       146040039       map_ratio=0.9994
chr1   146527987       147394444       -&gt;      chr1    147056425       147922330       map_ratio=0.9989
chr10  82045472        88931651        -&gt;      chr10   80285716        87171894        map_ratio=1.0000
chr11  43940000        46020000        -&gt;      chr11   43918450        45998449        map_ratio=1.0000
chr15  22805313        23094530        Fail    map_ratio=0.3607
chr15  22805313        28390339        -&gt;      chr15   22598414        28145193        map_ratio=0.8967
chr15  31080645        32462776        -&gt;      chr15   30788442        32170575        map_ratio=1.0000
chr15  72900171        78151253        -&gt;      chr15   72607830        77858911        map_ratio=1.0000
chr15  83219735        85722039        -&gt;      chr15   82550985        85178808        map_ratio=0.9800
chr16  15511655        16293689        -&gt;      chr16   15417798        16199832        map_ratio=1.0000
chr16  21950135        22431889        -&gt;      chr16   21938814        22420568        map_ratio=1.0000
chr16  28823196        29046783        -&gt;      chr16   28811875        29035462        map_ratio=1.0000
chr16  29650840        30200773        -&gt;      chr16   29639519        30189452        map_ratio=1.0000
chr17  1247834 1303556 -&gt;      chr17   1344540 1400262 map_ratio=1.0000
chr17  2496923 2588909 -&gt;      chr17   2593629 2685615 map_ratio=1.0000
chr17  16812771        20211017        -&gt;      chr17   16909457        20307704        map_ratio=1.0000
chr17  29107491        30265075        -&gt;      chr17   30780473        31938056        map_ratio=1.0000
chr17  34815904        36217432        Unmap
chr17  43705356        44164691        -&gt;      chr17   45627990        46087325        map_ratio=1.0000
chr2   50145643        51259674        -&gt;      chr2    49918505        51032536        map_ratio=1.0000
chr2   96742409        97677516        -&gt;      chr2    96076661        97011779        map_ratio=1.0000
chr2   111394040       112012649       -&gt;      chr2    110636463       111255072       map_ratio=1.0000
chr2   239716679       243199373       -&gt;      chr2    238808038       242183529       map_ratio=0.9622
chr22  19037332        21466726        Fail    map_ratio=0.8490
chr22  21920127        23653646        -&gt;      chr22   21565838        23311459        map_ratio=0.9996
chr22  51113070        51171640        -&gt;      chr22   50674642        50733212        map_ratio=1.0000
chr3   195720167       197354826       -&gt;      chr3    195993296       197627955       map_ratio=1.0000
chr4   1552030 2091303 -&gt;      chr4    1550303 2089576 map_ratio=1.0000
chr5   175720924       177052594       -&gt;      chr5    176293921       177625593       map_ratio=1.0000
chr7   72744915        74142892        -&gt;      chr7    73330912        74728554        map_ratio=0.9997
chr8   8098990 11872558        -&gt;      chr8    8241468 12015049        map_ratio=1.0000
chr9   140513444       140730578       -&gt;      chr9    137618992       137836126       map_ratio=1.0000
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ol class="arabic simple">
<li><p>Input BED file should have at least 3 columns (chrom, start, end). Additional columns will be kept as is.</p></li>
</ol>
</div>
</div>
</div>
<div class="section" id="compare-to-ucsc-liftover-tool">
<h1>Compare to UCSC liftover tool<a class="headerlink" href="#compare-to-ucsc-liftover-tool" title="Permalink to this headline">¶</a></h1>
<p>To access the accuracy of CrossMap, we randomly generated 10,000 genome intervals (download from <a class="reference external" href="https://sourceforge.net/projects/crossmap/files/hg19.rand.bed.gz/download">here</a>) with the
fixed interval size of 200 bp from hg19. Then we converted them into hg18 using CrossMap
and <a class="reference external" href="http://genome.ucsc.edu/cgi-bin/hgLiftOver">UCSC liftover tool</a> with default configurations. We compare CrossMap
to <a class="reference external" href="http://genome.ucsc.edu/cgi-bin/hgLiftOver">UCSC liftover tool</a> because it is the most widely
used tool to convert genome coordinates.</p>
<p>CrossMap failed to convert 613 intervals, and the UCSC liftover tool failed to convert 614
intervals. All failed intervals are exactly the same except for one region (chr2 90542908 90543108).
UCSC failed to convert it because this region needs to be split twice:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 29%" />
<col style="width: 30%" />
<col style="width: 40%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Original (hg19)</p></th>
<th class="head"><p>Split (hg19)</p></th>
<th class="head"><p>Target (hg18)</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>chr2 90542908  90543108 -</p></td>
<td><p>chr2 90542908 90542933 -</p></td>
<td><p>chr2    89906445        89906470 -</p></td>
</tr>
<tr class="row-odd"><td><p>chr2 90542908  90543108 -</p></td>
<td><p>chr2 90542933 90543001 -</p></td>
<td><p>chr2    87414583        87414651 -</p></td>
</tr>
<tr class="row-even"><td><p>chr2 90542908  90543108 -</p></td>
<td><p>chr2 90543010 90543108 -</p></td>
<td><p>chr2    87414276        87414374 -</p></td>
</tr>
</tbody>
</table>
<p>For genome intervals that were successfully converted to hg18, the start and end coordinates are
exactly the same between UCSC conversion and CrossMap conversion.</p>
<a class="reference internal image-reference" href="_images/CrossMap_vs_UCSC.png"><img alt="CrossMap_vs_UCSC_liftover.png" src="_images/CrossMap_vs_UCSC.png" style="width: 700.0px; height: 400.0px;" /></a>
</div>
<div class="section" id="citation">
<h1>Citation<a class="headerlink" href="#citation" title="Permalink to this headline">¶</a></h1>
<p><a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/24351709/">Zhao, H., Sun, Z., Wang, J., Huang, H., Kocher, J.-P., &amp; Wang, L. (2013). CrossMap: a versatile tool for coordinate conversion between genome assemblies. Bioinformatics (Oxford, England), btt730</a></p>
</div>
<div class="section" id="license">
<h1>LICENSE<a class="headerlink" href="#license" title="Permalink to this headline">¶</a></h1>
<p>CrossMap is distributed under <a class="reference external" href="http://www.gnu.org/copyleft/gpl.html">GNU General Public License</a></p>
<p>This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version. This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details. You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301 USA</p>
</div>
<div class="section" id="contact">
<h1>Contact<a class="headerlink" href="#contact" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p>Wang.Liguo AT mayo.edu</p></li>
</ul>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">What is CrossMap ?</a></li>
<li><a class="reference internal" href="#how-crossmap-works">How CrossMap works?</a></li>
<li><a class="reference internal" href="#release-history">Release history</a></li>
<li><a class="reference internal" href="#installation">Installation</a></li>
<li><a class="reference internal" href="#input-and-output">Input and Output</a><ul>
<li><a class="reference internal" href="#chain-file">Chain file</a></li>
<li><a class="reference internal" href="#user-input-file">User Input file</a></li>
<li><a class="reference internal" href="#output-file">Output file</a></li>
</ul>
</li>
<li><a class="reference internal" href="#usage">Usage</a><ul>
<li><a class="reference internal" href="#convert-bed-format-files">Convert BED format files</a></li>
<li><a class="reference internal" href="#convert-bam-cram-sam-format-files">Convert BAM/CRAM/SAM format files</a></li>
<li><a class="reference internal" href="#convert-wiggle-format-files">Convert Wiggle format files</a></li>
<li><a class="reference internal" href="#convert-bigwig-format-files">Convert BigWig format files</a></li>
<li><a class="reference internal" href="#convert-gff-gtf-format-files">Convert GFF/GTF format files</a></li>
<li><a class="reference internal" href="#convert-vcf-format-files">Convert VCF format files</a></li>
<li><a class="reference internal" href="#convert-maf-format-files">Convert MAF format files</a></li>
<li><a class="reference internal" href="#convert-gvcf-format-files">Convert GVCF format files</a></li>
<li><a class="reference internal" href="#convert-large-genomic-regions">Convert large genomic regions</a></li>
</ul>
</li>
<li><a class="reference internal" href="#compare-to-ucsc-liftover-tool">Compare to UCSC liftover tool</a></li>
<li><a class="reference internal" href="#citation">Citation</a></li>
<li><a class="reference internal" href="#license">LICENSE</a></li>
<li><a class="reference internal" href="#contact">Contact</a></li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="nav-item nav-item-0"><a href="#">CrossMap 0.5.2 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">What is CrossMap ?</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2020, Liguo Wang.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 3.1.2.
    </div>
  </body>
</html>